### DISCUSSION TITLE
Testing endpoints with middleware that relies on request headers

### DISCUSSION THREAD

**Discussion by mlms13** [RobinTail/express-zod-api#1799] [General]
Absolutely love `express-zod-api`, thanks so much for all the work you've put into it!

We encountered a bit of an interesting situation using the `testEndpoint` helper in our tests. At a high level:

- we have an endpoint
- the endpoint uses middleware to do API token auth
- the middleware checks for the presence of a specific header using `req.get('X-API-Token')`
- the tests created with `testEndpoint` always fail because `req.get is not a function`

For now, we've worked around this by providing our own implementation of Express's `request.get` method when setting up the `requestMock` for `testEndpoint`, like so:

```ts
testEndpoint({
  endpoint: ourEndpointToTest,
  requestMock: {
    method: 'POST',
    body: testBody,
    headers: testHeaders,
    get: implementationThatHopefullyMatchesExpressReqGet,
  }
});
```

I'm curious if there's a better way to do this, or if `express-zod-api` would consider adding a more robust base `requestMock` that more closely matches Express's own request object.

Happy to provide more detailed repro steps for how we got into the situation or more details about how we implemented our solution if that would be helpful. Thanks!

**Comment by RobinTail** [comment]
Hello @mlms13 ,

> Absolutely love express-zod-api, thanks so much for all the work you've put into it!

Thank you :)

> ~~requestMock: {~~

`requestProps` — that's correct

> the tests created with testEndpoint always fail because req.get is not a function
> we've worked around this by providing our own implementation of Express's request.get method 

This is not a workaround, this is exactly how it's designed to be.

> would consider adding a more robust base 
> that more closely matches Express's own request object

I did consider that but I could not find a suitable solution. In case you know one, please let me know and I'll try it.

**Reply by RobinTail** [reply]
https://www.npmjs.com/package/node-mocks-http

Perhaps this one can be utilized. It seems to be abstract enough.

**Reply by RobinTail** [reply]
@mlms13 , I'm going to try to adopt that library, but it would be breaking for `testEndpoint()`, therefore it will be a feature of v20

**Reply by RobinTail** [reply]
https://github.com/RobinTail/express-zod-api/pull/1800 , @mlms13

**Comment by RobinTail** [comment]
Dear @mlms13 ,

Please check out the version `20.0.0-beta.1` released today.
It has a number of changes, though I made a script to make the migration easier — read more [in Changelog](https://github.com/RobinTail/express-zod-api/blob/prep-v20/CHANGELOG.md#v2000).

Amongst other improvements, there is new testing approach that is completely detached from the mocks of your testing frameworks, moreover it offers a completely operational mocks for request and response, enabling its usage outta box without any tricks, thus eliminating the confusion.

Please let me know how it works.

**Reply by RobinTail** [reply]
Still need your feedback, @mlms13

**Reply by RobinTail** [reply]
@mlms13 , did you have a chance to try `v20.0.0-beta.1` ?

**Reply by mlms13** [reply]
Sorry for the slow feedback! We're a few versions behind, so it may be a bit before we're able to update to the v20 beta, but I'll try to get to this soon!

**Reply by RobinTail** [reply]
v20.0.0 released 🚀
